src/shared/metrics.h

```diff
@@ -0,0 +1,133 @@
+                         * SD_BUS_VTABLE_END in a way that is both compatible with pedantic versions of C and
+                         * C++. It's unused otherwise. */
+                        size_t _reserved;
+                } end;
+                struct {
+                        const char *name;
+                        const char *description;
+                        MetricType type;
+                        metric_family_get_t get;
+                } metric_family;
+        } x;
+};
+
+#define IO_SYSTEMD_MANAGER "io.systemd.Manager."
+#define UNIT_STATES_TOTAL "unit_states_total"
+#define UNIT_TYPES_TOTAL "unit_types_total"
+#define UNIT_STATE "unit_state"
+#define IO_SYSTEMD_MANAGER_UNIT_STATES_TOTAL IO_SYSTEMD_MANAGER UNIT_STATES_TOTAL
+#define IO_SYSTEMD_MANAGER_UNIT_TYPES_TOTAL IO_SYSTEMD_MANAGER UNIT_TYPES_TOTAL
+#define IO_SYSTEMD_MANAGER_UNIT_STATE IO_SYSTEMD_MANAGER UNIT_STATE
+
+#define METRICS_VTABLE_START()                                          \
+        {                                                               \
+                .type = _METRICS_VTABLE_START,                          \
+                .x = {                                                  \
+                        .start = {                                      \
+                                .element_size = sizeof(metrics_vtable), \
+                        },                                              \
+                },                                                      \
+        }
+
+#define METRICS_FAMILY(_name, _description, _type, _get)                \
+        {                                                               \
+                .type = _METRICS_FAMILY,                                \
+                .x = {                                                  \
+                        .metric_family = {                              \
+                                .name = _name,                          \
+                                .description = _description,            \
+                                .type = _type,                          \
+                                .get = _get,                            \
+                        },                                              \
+                },                                                      \
+        }
+
+#define METRICS_VTABLE_END                                              \
+        {                                                               \
+                .type = _METRICS_VTABLE_END,                            \
+                .x = {                                                  \
+                        .end = {                                        \
+                                ._reserved = 0,                         \
+                        },                                              \
+                },                                                      \
+        }
```

Comment:
passing feedback from @Daan. THe vtable staff can be significantly simplified by reducing it to a simple array of structs with name, describe, type, and callback.